crates/ide_db/src/helpers.rs

```diff
@@ -81,16 +89,28 @@ pub fn try_resolve_derive_input(
     attr: &ast::Attr,
     cursor: &ast::Ident,
 ) -> Option<PathResolution> {
-    let path = get_path_in_derive_attr(sema, attr, cursor)?;
-    let scope = sema.scope(attr.syntax());
-    // FIXME: This double resolve shouldn't be necessary
-    // It's only here so we prefer macros over other namespaces
-    match scope.speculative_resolve_as_mac(&path) {
-        Some(mac) if mac.kind() == hir::MacroKind::Derive => Some(PathResolution::Macro(mac)),
-        Some(_) => return None,
-        None => scope
+    let derives = sema.resolve_derive_macro(attr)?;
+
+    let tt = attr.token_tree()?;
+    if !tt.syntax().text_range().contains_range(cursor.syntax().text_range()) {
+        return None;
+    }
+    let (path, is_qualifier) = get_path_at_cursor_in_tt(cursor)?;
+
+    if is_qualifier {
+        sema.scope(attr.syntax())
             .speculative_resolve(&path)
-            .filter(|res| matches!(res, PathResolution::Def(ModuleDef::Module(_)))),
+            .filter(|res| matches!(res, PathResolution::Def(ModuleDef::Module(_))))
+    } else {
+        let cursor = cursor.syntax();
+        let idx = tt
+            .syntax()
+            .children_with_tokens()
+            .filter_map(SyntaxElement::into_token)
+            .take_while(|tok| tok != cursor)
+            .filter(|t| t.kind() == T![,])
+            .count();
+        derives.get(idx)?.map(PathResolution::Macro)
```

Comment:
Let's move this indexing logic into semantics, such that API for IDE is "here's a $Token inside $Derive, what it resolves to?". That's the general shape of the API of semantics, which uses "syntax trees with identity" for `resolve_` familiy of methods. 

The difference here is that typically resolve methods take some well-defined syntactics things, like `resolve_extern_crate(ast::ExternCrate)`, while here we have to use a rather amorphous `resolve_derive_trait(ast::Ident)`.

Actually, I think we can make the API a touch more intuitive like this: 

```rust
impl Semantics {
    /// Invariant: `derive` is the parrent attr of `ident`. 
    /// For `#[dervie(serde::Serialize)]`, `dervie` points at the whole `#[...]`, while
    /// `ident` can point at either of `serde` or `Serialize`. 
    fn resolve_derive_trait(&self, derive: &ast::Attr, ident: &ast::Ident)
}
```

Reply:
Ye I just read your reply on the original issue as well. That should make this a bunch nicer.